Tedefo 4805 efx rules translator

rousso opened the pull request:
This adds a transpiler to generate Schematron from EFX rules files.

rousso committed:
Add EFX Rules to Schematron transpiler foundation

Infrastructure:
- Add EfxRulesTranslator interface with translateRules methods
- Add factory methods to EfxTranslatorFactory
- Add convenience methods to EfxTranslator
- Update eforms-core dependency to 1.6.0-SNAPSHOT

Data Model:
- Create Schematron model classes (SchematronSchema, SchematronPattern,
  SchematronRule, SchematronAssert, SchematronLet, SchematronPhase,
  SchematronFile)
- Add Freemarker templates for Schematron XML generation
  (complete-validation.ftl, pattern.ftl, rule.ftl, assert.ftl, let.ftl)

Components:
- Create SchematronMarkupGenerator with Freemarker integration
- Create EfxRulesTranslatorV2 skeleton extending EfxExpressionTranslatorV2
- Add Freemarker 2.3.31 dependency to pom.xml

Status: Foundation complete, listener methods TODO

rousso committed:
Fix license text in multiple interface files

rousso committed:
Added notice subtype handling to SdkSymbolResolver.

rousso committed:
Add EFX Rules to Schematron translator

rousso committed:
Refactor: Rename ValidatorMarkupGenerator to ValidatorGenerator and SchematronMarkupGenerator to SchematronGenerator

bertrand-lorentz commented on pom.xml:
```diff
@@ -48,11 +48,12 @@
     <!-- Versions - Third-party libraries -->
     <version.antlr4>4.13.1</version.antlr4>
     <version.commons-lang3>3.18.0</version.commons-lang3>
+    <version.freemarker>2.3.31</version.freemarker>
```
This version dates back to 2021. The latest version is 2.3.34, from 2024, and its requirements are the same. So it would be better to use that version.

rousso replied:
Upgraded to 2.3.34. All 654 tests pass.

bertrand-lorentz committed:
EfxRulesTranslatorV2Test: Fix unit tests on Windows

The "schematrons.json" were indicated as different when running on
Windows.
Stripping the trailing whitespace avoids this problem.

bertrand-lorentz commented on src/test/resources/eu/europa/ted/efx/sdk2/EfxRulesTranslatorV2Test/testAssertAndReport_Simple/dynamic/validation-stage-1a-1.sch:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<pattern id="EFORMS-validation-stage-1a-1" xmlns="http://purl.oclc.org/dsdl/schematron">
+    <rule context="/*/PathNode/TextField">
+        <assert id="R-K7P-M2Q" role="error" test=".">rule|text|R-K7P-M2Q</assert>
```
Values in the "role" attribute are currently in uppercase: role="ERROR"

rousso replied:
Changed to uppercase.

bertrand-lorentz commented on src/main/java/eu/europa/ted/eforms/sdk/schematron/SchematronAssert.java:
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2022 European Union
```
The copyright year should be 2025 or 2026.
I guess the same applies to all added copyright headers.

rousso replied:
The copyright year should be the year when the file was added to the repo.

rousso replied:
Copyright years reflect the creation date of the file and remain unchanged throughout the file's lifetime.

rousso replied:
Will update all copyright headers to 2025 for the newly added files in the `schematron` package.

bertrand-lorentz commented on src/test/resources/eu/europa/ted/efx/sdk2/EfxRulesTranslatorV2Test/testDiagnostics_NoDuplicateEntries/input.efx:
```diff
@@ -0,0 +1,16 @@
+// Test: Same field used in multiple rules
+// Verifies no duplicate diagnostics when same field referenced twice
```
This file does not cause any diagnostics attribute to be present in the schematron, so it does not really test what is described here.

rousso replied:
Fixed. The test now uses ND-Root as context with BT-00-Text as subject, which properly generates the diagnostics attribute for testing duplicate detection.

bertrand-lorentz commented on src/test/resources/eu/europa/ted/efx/sdk2/EfxRulesTranslatorV2Test/testWhen_SimpleCondition/static/validation-stage-1a-1.sch:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<pattern id="EFORMS-validation-stage-1a-1" xmlns="http://purl.oclc.org/dsdl/schematron">
+    <rule context="/*/SubNode">
+        <assert id="R-K7P-M2Q" role="error" diagnostics="ND-SubNode_BT-00-Text" test="(not(../PathNode/TextField)) or (../PathNode/TextField/normalize-space(text()) = 'open')">rule|text|R-K7P-M2Q</assert>
```
In current schematrons, the test is the other way around, with the condition at the end:
`(../PathNode/TextField/normalize-space(text()) = 'open') or not(../PathNode/TextField)`

Keeping this same order would help people familiar with the current schematrons.
And it might make validation a little faster for notices that are valid: only the left-hand side of the "or" clause would need to be evaluated.

rousso replied:
Done in 95c1176. Changed the test expression order for WHEN clauses to put the assertion first and the condition check second: `(expression) or (not(condition))` instead of `(not(condition)) or (expression)`.

This affects:
- exitWhenClause
- exitAssertClause  
- exitReportClause
- exitOtherwiseAssertClause
- exitOtherwiseReportClause
- exitFallbackRule

All tests pass with the new ordering.

bertrand-lorentz commented on src/test/resources/eu/europa/ted/efx/sdk2/EfxRulesTranslatorV2Test/testVariable_Global_AppearsBeforeIncludes/static/complete-validation.sch:
```diff
@@ -0,0 +1,36 @@
+    <ns prefix="fn" uri="http://www.w3.org/2005/xpath-functions" />
+
+    <let name="sdkVersion" value="&quot;2.0.0&quot;"/>
+    <let name="noticeType" value="&quot;16&quot;"/>
```
The value seems incorrect, I don't think it should have &quot;

rousso replied:
The `&quot;` entities are actually necessary and correct here. 

EFX supports both single and double-quoted string literals:
- `LET text : $foo = "it's a test";`
- `LET text : $bar = 'say "hello"';`

We pass through the quote type from the EFX source to XPath to handle strings containing the opposite quote character. If we blindly converted all to single quotes, strings like `"it's a test"` would break because the internal single quote would terminate the string.

The XML entities (`&quot;`) are just how XML escapes double quotes in attribute values - the XPath processor sees the correct string literal with its original quotes.

bertrand-lorentz commented on src/test/resources/eu/europa/ted/efx/sdk2/EfxRulesTranslatorV2Test/testWhen_WithOtherwise/static/validation-stage-1a-2.sch:
```diff
@@ -0,0 +1,8 @@
+    <rule context="/*/PathNode/TextField">
+        <report id="R-X3F-N8W" role="info" test="(not(./normalize-space(text()) = 'restricted')) or (../NumberField/number() &gt; 0)">rule|text|R-X3F-N8W</report>
+        <assert id="R-H9T-V5L" role="warning" test="(not(./normalize-space(text()) = 'negotiated')) or (../IndicatorField)">rule|text|R-H9T-V5L</assert>
+        <report id="R-B6J-C4R" role="info" test="(not(not(./normalize-space(text()) = 'open') and not(./normalize-space(text()) = 'restricted') and not(./normalize-space(text()) = 'negotiated'))) or (./normalize-space(text()) != '')">rule|text|R-B6J-C4R</report>
```
I would not necessarily expect the condition on 'open' to be here: it does not apply to subtype 2, so having it in the "report" element corresponding to the OTHERWISE might not be expected.
It can stay like it is, but it will need to be clearly indicated for the documentation on "OTHERWISE".

rousso replied:
You're right - this is a subtlety in how OTHERWISE works. 

The OTHERWISE condition includes ALL preceding WHEN conditions from the same WITH block, regardless of their IN clauses. In this case:
- `WHEN text == 'open' FOR IN 1` 
- `WHEN text == 'restricted' FOR IN 1, 2`
- `WHEN text == 'negotiated' FOR IN 2`
- `OTHERWISE REPORT ... FOR IN 2`

The OTHERWISE for notice type 2 checks: `NOT ('open' OR 'restricted' OR 'negotiated')` even though the 'open' condition only applies to type 1.

This is safe but not optimal - it includes a redundant check. We can document this behavior clearly in the OTHERWISE documentation to explain that it combines all WHEN conditions from the WITH block, not just those that apply to the same notice types.

The alternative would be to filter the inverted conditions by notice type, but that would complicate the logic and the current behavior is correct (just not minimal).

Also, the user needs to create a rule that makes sense. If the user adds an otherwise clause which contradicts notice filtering, then the input may not make sense. We can add an error for this case later, but I found it an overkill since we do not yet use this feature at all. In any case OTHERWISE should run ONLY IF none of the WHEN conditions was met regardless of notice filtering.

bertrand-lorentz replied:
OK, this can stay as it is.
I'm not marking this as resolved, just to keep the explanations visible by default.

bertrand-lorentz commented on src/main/java/eu/europa/ted/efx/interfaces/ValidatorGenerator.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    Map<String, String> generateOutput(
+        CompleteValidation completeValidation
+    ) throws IOException;
```
IOException might not be the best suited exception: there's no I/O implied by this method's definition (it reads from the parameter and returns strings).

rousso replied:
Fixed by removing IOException from the interface. The implementation now wraps IOException in RuntimeException since this method performs in-memory transformations without actual I/O operations.

rousso committed:
Remove IOException from ValidatorGenerator interface

The method transforms an in-memory model to strings with no I/O operations implied by the signature. IOException is an implementation detail of Freemarker template loading, which is now handled internally by wrapping in RuntimeException.

rousso committed:
Upgrade Freemarker from 2.3.31 to 2.3.34

rousso commented on src/main/java/eu/europa/ted/efx/sdk2/EfxRulesTranslatorV2.java:
```diff
@@ -0,0 +1,751 @@
+   * Extends ExpressionPreprocessor to inherit common late-bound expression handling
+   * and adds Rules-specific variable declaration processing.
+   */
+  class RulesPreprocessor extends ExpressionPreprocessor {
```
Won't do. Heavy preprocessor refactoring will be done after efx-2 syntax has been fully defined.

rousso committed:
Use uppercase for Schematron role attribute

rousso committed:
Fix testDiagnostics_NoDuplicateEntries to actually generate diagnostics

rousso committed:
Swap test expression order in WHEN clauses - put assertion before condition

rousso commented on src/main/java/eu/europa/ted/efx/model/rules/CompleteValidation.java:
```diff
@@ -0,0 +1,91 @@
+        boolean aNum = isNumeric(a);
+        boolean bNum = isNumeric(b);
+        if (aNum && bNum) {
+            return Integer.compare(Integer.parseInt(a), Integer.parseInt(b));
```
This is safe. The `isNumeric()` method on line 83 already calls `Integer.parseInt()` and catches `NumberFormatException`, returning false if parsing fails. The `parseInt()` on line 78 is only reached when both `aNum` and `bNum` are true (line 77), meaning both strings have already been successfully parsed by `isNumeric()`. Therefore, the `parseInt()` calls on line 78 are guaranteed to succeed.

rousso commented on src/main/java/eu/europa/ted/efx/model/rules/CompleteValidation.java:
```diff
@@ -0,0 +1,91 @@
+        boolean aNum = isNumeric(a);
+        boolean bNum = isNumeric(b);
+        if (aNum && bNum) {
+            return Integer.compare(Integer.parseInt(a), Integer.parseInt(b));
```
This is safe. The `isNumeric()` method on line 83 already calls `Integer.parseInt()` and catches `NumberFormatException`, returning false if parsing fails. The `parseInt()` on line 78 is only reached when both `aNum` and `bNum` are true (line 77), meaning both strings have already been successfully parsed by `isNumeric()`. Therefore, the `parseInt()` calls on line 78 are guaranteed to succeed.

bertrand-lorentz committed:
Fix build with JDK 11

Stream.toList() was only added in JDK 16, so instead use
"collect(Collectors.toList())"
